curvefs/src/metaserver/copyset/copyset_node.cpp

```diff
@@ -0,0 +1,509 @@
+}
+
+class OnSnapshotSaveDoneImpl : public OnSnapshotSaveDone {
+ public:
+    OnSnapshotSaveDoneImpl(CopysetNode* node, braft::SnapshotWriter* writer,
+                           braft::Closure* snapDone, uint64_t startTimeUs)
+        : node_(node),
+          writer_(writer),
+          snapDone_(snapDone),
+          statusCode_(MetaStatusCode::OK),
+          startTimeUs_(startTimeUs) {}
+
+    void Run() override {
+        std::unique_ptr<OnSnapshotSaveDoneImpl> selfGuard(this);
+        brpc::ClosureGuard doneGuard(snapDone_);
+
+        if (statusCode_ != MetaStatusCode::OK) {
+            RaftSnapshotMetric::GetInstance().OnSnapshotSaveFailure();
+            return;
+        }
+
+        writer_->add_file(kMetaDataFilename);
+        RaftSnapshotMetric::GetInstance().OnSnapshotSaveComplete(
+            TimeUtility::GetTimeofDayUs() - startTimeUs_);
+    }
+
+    void SetSuccess() override {
+        LOG(INFO) << "Copyset " << node_->Name() << " save snapshot success";
+    }
+
+    void SetError(MetaStatusCode code) override {
+        statusCode_ = code;
+        snapDone_->status().set_error(statusCode_,
+                                      "Save metadata failed, err: %s",
+                                      MetaStatusCode_Name(statusCode_).c_str());
+        LOG(ERROR) << "Copyset " << node_->Name()
+                   << " save snapshot failed, err: "
+                   << MetaStatusCode_Name(statusCode_);
+    }
+
+ private:
+    CopysetNode* node_;
+    braft::SnapshotWriter* writer_;
+    braft::Closure* snapDone_;
+    MetaStatusCode statusCode_;
+    uint64_t startTimeUs_;
+};
+
+void CopysetNode::on_snapshot_save(braft::SnapshotWriter* writer,
+                                   braft::Closure* done) {
+    LOG(INFO) << "Copyset " << name_ << " saving snapshot to '"
+              << writer->get_path() << "'";
+
+    brpc::ClosureGuard doneGuard(done);
+    auto startTimeUs = TimeUtility::GetTimeofDayUs();
+
+    // flush all flying operators
+    applyQueue_->Flush();
+
+    // save conf
+    std::string confFile = writer->get_path() + "/" + kConfEpochFilename;
+    if (0 != SaveConfEpoch(confFile)) {
+        LOG(ERROR) << "Copyset " << name_
+                   << " save snapshot failed, save epoch file failed";
+        done->status().set_error(errno, "Save conf file failed, error = %s",
+                                 strerror(errno));
+        return;
+    }
+
+    writer->add_file(kConfEpochFilename);
+
+    std::string metaDataFile = writer->get_path() + "/" + kMetaDataFilename;
+
+    // TODO(wuhanqing): MetaStore::Save will start a thread and do task
+    // asynchronously, after task completed it will call
+    // OnSnapshotSaveDoneImpl::Run
+    // BUT, this manner is not so clear, maybe it better to make thing
+    // asynchronous directly in here
+    metaStore_->Save(metaDataFile, new OnSnapshotSaveDoneImpl(
+                                       this, writer, done, startTimeUs));
```

Comment:
`done->release()` return value type is `google::protobuf::Closure*`, but OnSnapshotSaveDoneImpl accepts `braft::Closure*`.
so I just pass done, and release in the below, otherwise, i have to add a type cast, like `statis_cast<braft::Closure*>(done->release())`